pyvo/utils/tests/test_download.py

```diff
@@ -0,0 +1,194 @@
+    # dependencies of the latter, so it's enough to handle them with one variable
+    import boto3
+    from botocore.exceptions import ClientError
+    from moto import mock_s3
+    HAS_MOTO = True
+except ImportError:
+    HAS_MOTO = False
+
+# Useful variables
+# For http_download:
+get_pkg_data_contents = partial(
+    get_pkg_data_contents, package=__package__, encoding='binary')
+data_re = re.compile('http://example.com/data.*')
+# For aws_download:
+s3_bucket = 'pyvo-bucket'
+s3_key = 'key1/key2/somekey.txt'
+
+
+class ContextAdapter(requests_mock.Adapter):
+    """
+    requests_mock adapter where ``register_uri`` returns a context manager
+    """
+    @contextmanager
+    def register_uri(self, *args, **kwargs):
+        matcher = super().register_uri(*args, **kwargs)
+
+        yield matcher
+
+        self.remove_matcher(matcher)
+
+    def remove_matcher(self, matcher):
+        if matcher in self._matchers:
+            self._matchers.remove(matcher)
+
+
+@pytest.fixture(scope='function')
+def mocker():
+    with requests_mock.Mocker(
+        adapter=ContextAdapter(case_sensitive=True)
+    ) as mocker_ins:
+        yield mocker_ins
+
+
+def test__filename_from_url():
+    urls = [
+        'https://example.com/files/myfile.pdf?user_id=123',
+        'https://example.com/files/myfile.pdf',
+        'http://somesite.com/service?file=/location/myfile.pdf&size=large'
+    ]
+
+    for url in urls:
+        filename = _filename_from_url(url)
+        assert filename == 'myfile.pdf'
+
+
+@pytest.fixture(name='http_mock')
+def _data_downloader(mocker):
+    def callback(request, context):
+        fname = request.path.split('/')[-1]
+        return get_pkg_data_contents(f'data/{fname}')
+
+    with mocker.register_uri(
+        'GET', data_re, content=callback, headers={'content-length': '901'},
+    ) as matcher:
+        yield matcher
+
+
+def test_http_download__noPath(http_mock):
+    filename = http_download('http://example.com/data/basic.xml',
+                             local_filepath=None, cache=False)
+    assert filename == 'basic.xml'
+    os.remove('basic.xml')
+
+
+def test_http_download__noFile(http_mock):
+    with pytest.raises(urllib.error.URLError):
+        http_download('http://example.com/data/nofile.fits')
+
+
+def test_http_download__wPath(http_mock):
+    filename = http_download('http://example.com/data/basic.xml',
+                             local_filepath='basic2.xml', cache=False)
+    assert filename == 'basic2.xml'
+    assert os.path.exists('basic2.xml')
+    os.remove('basic2.xml')
+
+
+def test_http_download__wrong_cache(http_mock):
+    # get the file first
+    with open('basic.xml', 'w') as fp:
+        fp.write('some content')
+    # get it from cache
+    with pytest.warns(PyvoUserWarning):
+        http_download('http://example.com/data/basic.xml',
+                      local_filepath='basic.xml', cache=True)
+    assert os.path.getsize('basic.xml') == 901
```

Comment:
I suppose some content related test is needed instead of size, to make windows happy.